`EIP-7823` and `EIP-7883` implementation --- Osaka `MODEXP` changes

OlivierBBB opened the pull request:
<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Implements fork-aware MODEXP per Osaka (EIP‑7823), introducing London/Osaka MODEXP metadata, pricing and XBS OOB calls, updating MMU and counters to fork-specific sizes/costs, wiring fork into ZkCounter via RPC, and adjusting tests.
> 
> - **zktracer (core)**:
>   - Make `ZkCounter` fork-aware; derive `OpCodes`/`Trace` from `fork` and use fork in MODEXP logic and line counts.
>   - Split MODEXP into forked implementations: `LondonModexpSubsection`/`OsakaModexpSubsection`, with `London/OsakaModexpMetadata`.
>   - Add forked MODEXP ops: `London/OsakaBlakeModexpDataOperation`, `London/OsakaModexpPricingOobCall`, `London/OsakaModexpXbsOobCall`.
>   - Update exponent-log and return sizing to use fork-specific multipliers and max input sizes.
> - **MMU**:
>   - Pass `fork` into `Mmu`/`MmuInstructions`; adjust `ModexpData`/`ModexpZero` micro-rows and component sizes per fork.
> - **Precompile processing**:
>   - Route PRC_MODEXP to London/Osaka subsections; count rows via `ModexpMetadata.getNumberOfRowsForModexp()`; make MODEXP success depend on forked bounds and pricing.
>   - Use fork-specific MODEXP rows and Blake rows via unified `BlakeModexpData` API.
> - **RPC / Counter**:
>   - `GenerateLineCountsV2` and `ConflatedCountTracesV2`: construct `ZkCounter` with resolved `fork`.
> - **Tests**:
>   - Update unit/integration tests to reflect fork-aware MODEXP limits, pricing, gas, and helper utilities.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2fedec6511a849ce7dc352259ad6d03f5f55436f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

OlivierBBB committed:
feat: constraints update

OlivierBBB committed:
constraints update

OlivierBBB committed:
first commit

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -754,8 +754,8 @@ public static MmuCall forModexpExtractEbs(
-  public static MmuCall forModexpExtractMbs(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata metaData) {
+  public static MmuCall extractMbsForModexp(
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata metaData) {
```
Is it normal ? What's happening for OsakaModexpSection ?

letypequividelespoubelles replied:
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/precompiles/modexpMetadata/LondonModexpMetadata.java:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  public int getMaxInputSize() {
+    return 512;
```
this constant has a name somewhere ...

OlivierBBB replied:
It doesn't in the HUB constraints.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/OsakaModexpSubsection.java:
```diff
@@ -0,0 +1,24 @@
+  public OsakaModexpSubsection(Hub hub, CallSection callSection) {
+    super(hub, callSection);
+  }
+}
```
this comment seems legit

OlivierBBB committed:
various ...

OlivierBBB committed:
stuff ...

OlivierBBB committed:
clean up

With Lorenzo and François

OlivierBBB committed:
fix: 99% up to par + @EqualsAndHash(callSuper = true)

One failing test (MODEXP limits)

OlivierBBB committed:
fix: ModexpTests

the old tests were overwriting mbs in memory !

OlivierBBB committed:
feat: fix ZkCounter event detection

OlivierBBB committed:
spotless

OlivierBBB committed:
spotless ... again ?

OlivierBBB committed:
feat: Osaka MODEXP pricing OOB call

OlivierBBB committed:
stuff

OlivierBBB committed:
spotless

OlivierBBB committed:
setting TOT in MODEXP's MMU instructions (DATA and ZERO)

OlivierBBB committed:
small stuff

OlivierBBB committed:
bug: use OsakaBlakeModexpOperation for Osaka rather than LondonBla

OlivierBBB committed:
feat: most tests are passing

OlivierBBB committed:
feat: constraints update

OlivierBBB committed:
merge arith-dev

OlivierBBB committed:
constraints update

OlivierBBB committed:
feat: byteMultiplier of leadLog in MODEXP made fork dependent

leadLog is essentially defined as

   log_2(exponent_leading_word) + multiplier * (ebs - 32)

with multiplier = 8 pre-Osaka, and 16 post-Osaka

OlivierBBB committed:
feat: ugly test fixes

OlivierBBB committed:
feat: fork dependent MODEXP_PRICING byte multiplier m * (ebs - 32)

OlivierBBB committed:
feat: fork dependent xbsNonzero for MODEXP_XBS

OlivierBBB committed:
ras: constraints commit update

OlivierBBB committed:
feat: add fork field to ZkCounter class

OlivierBBB committed:
spotless

OlivierBBB committed:
fix: use fork appropriate gasLimit in LowGasStipendPrecompileCallTests

OlivierBBB committed:
Merge branch 'arith-dev' into EIP-7823-and-7883---Osaka-MODEXP-changes

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeComponents.java:
```diff
@@ -15,8 +15,8 @@
 
-import static net.consensys.linea.zktracer.module.blake2fmodexpdata.BlakeModexpDataOperation.BLAKE2f_HASH_INPUT_OFFSET;
-import static net.consensys.linea.zktracer.module.blake2fmodexpdata.BlakeModexpDataOperation.BLAKE2f_HASH_INPUT_SIZE;
+import static net.consensys.linea.zktracer.module.blake2fmodexpdata.BlakeModexpOperation.BLAKE2f_HASH_INPUT_OFFSET;
```
why rm the data ? We have it for other modules (EcData, etc ...)

OlivierBBB replied:
Changed it back.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpData.java:
```diff
@@ -52,21 +52,26 @@ public ModuleName moduleKey() {
   }
 
-  private void callWcpForIdCheck(final int operationID) {
+  private void callWcpForIdCheck(final long operationID) {
```
operation is from hubStamp, whicj is a i32, so int is enough no ?

OlivierBBB replied:
It's enough, I changed it because at other points the `operationId` is required an `long` and it was complaining.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpDataOperation.java:
```diff
@@ -178,4 +189,18 @@ private void commonTrace(
+    return blake2fComponents.isPresent();
+  }
+
+  public static int legalModexpComponentByteSize(Fork fork) {
```
should add a commet to say that legal is accepted by linea in Prague vs non exceptionnal in Osaka

OlivierBBB replied:
Done

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/LondonBlakeModexpData.java:
LondonBlakeModexp == OsakaBmakeModexp as it's the underlying operation which is different depeding on the Fork. So Imho, it seems no need to have an abstract BlakeModexpData and two extends. Only keep a non abstract BlakeModexppData seems enough.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -731,8 +731,8 @@ public static MmuCall partialCopyOfReturnDataforBlake(
-  public static MmuCall forModexpExtractBbs(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata metaData) {
+  public static MmuCall extractBbsForModexp(
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata metaData) {
```
why sopecialized a LondonModexpSection, and not only an abstract ModexpSubsection ? 
As it is, it seems we only accept London (but as Osaka extends London it works), so it's strange.

OlivierBBB replied:
There is no `ModexpSubsection`.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -742,8 +742,8 @@ public static MmuCall forModexpExtractBbs(
-  public static MmuCall forModexpExtractEbs(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata metaData) {
+  public static MmuCall extractEbsForModexp(
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata metaData) {
```
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -767,7 +767,7 @@ public static MmuCall forModexpExtractMbs(
 
   public static MmuCall forModexpLoadLead(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata metaData) {
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata metaData) {
```
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -776,14 +776,14 @@ public static MmuCall forModexpLoadLead(
-  public static MmuCall forModexpExtractBase(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
+  public static MmuCall extractModexpBase(
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
```
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -798,14 +798,14 @@ public static MmuCall forModexpExtractBase(
-  public static MmuCall forModexpExtractExponent(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
+  public static MmuCall extractModexpExponent(
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
```
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -820,13 +820,13 @@ public static MmuCall forModexpExtractExponent(
-  public static MmuCall forModexpExtractModulus(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
+  public static MmuCall extractModexpModulus(
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
```
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -837,29 +837,29 @@ public static MmuCall forModexpExtractModulus(
 
   public static MmuCall forModexpFullResultCopy(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
```
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -837,29 +837,29 @@ public static MmuCall forModexpExtractModulus(
 
   public static MmuCall forModexpPartialResultCopy(
-      final Hub hub, final ModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
+      final Hub hub, final LondonModexpSubsection subsection, final ModexpMetadata modExpMetadata) {
```
same

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/oob/precompiles/modexp/pricingOobCall/OsakaModexpPricingOobCall.java:
```diff
@@ -0,0 +1,99 @@
+import net.consensys.linea.zktracer.module.wcp.Wcp;
+import org.apache.tuweni.bytes.Bytes;
+
+public class OsakaModexpPricingOobCall extends LondonModexpPricingOobCall {
```
could extend the abstract as well ... there is noth,g common between the London and Osaka.

But it's purely stylistic comment

OlivierBBB replied:
I don't have an opinion on this.

amkCha replied:
If you change it though, you might want to check all the other places where it's casting as `LondonModexpPricingOobCall` because `OsakaModexpPricingOobCall` won't be able to be cast to a london one

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/precompiles/modexpMetadata/OsakaModexpMetadata.java:
```diff
@@ -0,0 +1,59 @@
+
+  @Override
+  public int getForkAppropriateLeadLogByteMultiplier() {
+    return 16;
```
doesn't have a name in the constant ?

OlivierBBB replied:
> doesn't have a name in the constant ?

It's 16 in the EIP, and 16 in the lisp (and used to be 8 in the Yellow paper). I do agree that a name would make sense. Maybe `MODEXP_EXPONENT___BYTE_WEIGHT_IN_LOG` or so.

OlivierBBB replied:
https://github.com/Consensys/linea-specification/issues/325

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/OsakaModexpSubsection.java:
```diff
@@ -0,0 +1,65 @@
+    super(hub, callSection, modexpMetadata);
+    checkState(
+        modexpMetadata instanceof OsakaModexpMetadata,
+        "modexpMetadata must be LondonModexpMetadata");
```
cursor is right

OlivierBBB replied:
Done

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -341,9 +341,18 @@ public List<Module> getTracelessModules() {
-          wcp, modexpEffectiveCall, modexpLargeCall, blakeEffectiveCall, blakeRounds);
+
+  @Getter
+  private final BlakeModexp blakeModexp =
```
I think we don't need to have a fork defendant module, see previous comment

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -374,7 +383,7 @@ public List<Module> realModule() {
         add,
         bin,
-        blakeModexpData,
+        blakeModexp,
```
why rename ? It's still *data in the spec no ? + now it doesn't match the key of the module.

tbf, i would keep the previous name ...

OlivierBBB replied:
I switched names back

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -448,7 +457,7 @@ public Hub(final ChainConfig chain, Map<Long, Hash> historicalBlockHashes) {
     rlpAddr = new RlpAddr(this, trm, keccak);
     blockdata = setBlockData(this, wcp, euc, chain);
-    mmu = new Mmu(euc, wcp);
+    mmu = new Mmu(euc, wcp, fork);
```
we should refactor the mmuCall to give as an input only the ModexMetaData ...

In a next cleaning PR

letypequividelespoubelles commented on arithmetization/src/test/java/net/consensys/linea/zktracer/module/limits/precompileLimits/ModexpLimitsTests.java:
```diff
@@ -125,6 +157,10 @@ private static Stream<Arguments> modexpInput() {
       }
     }
+
+    // arguments.clear();
```
?

OlivierBBB replied:
removed

letypequividelespoubelles commented on arithmetization/src/test/resources/junit-platform.properties:
```diff
@@ -1,3 +1,3 @@
-# junit.jupiter.execution.parallel.enabled = true
-# junit.jupiter.execution.parallel.mode.default = concurrent
-# junit.jupiter.execution.parallel.config.strategy = dynamic
+junit.jupiter.execution.parallel.enabled = true
```
wut ?

OlivierBBB replied:
Included by mistake

letypequividelespoubelles commented on testing/src/main/java/net/consensys/linea/reporting/TracerTestBase.java:
```diff
@@ -30,7 +30,7 @@ public class TracerTestBase {
     // Configure chain information and fork before any tests are run, including any methods used as
     // MethodSource.
-    TracerTestBase.chainConfig = ChainConfig.MAINNET_TESTCONFIG(getForkOrDefault(OSAKA));
+    TracerTestBase.chainConfig = ChainConfig.MAINNET_TESTCONFIG(getForkOrDefault(PRAGUE));
```
back to osaka, the release will be for Osaka

OlivierBBB replied:
Done

letypequividelespoubelles commented on plan.md:
to delete

OlivierBBB replied:
Done

OlivierBBB committed:
fix: disable parallelism (enabled by mistake)

OlivierBBB committed:
wip: fix of huge gas consumption in 'preCallProgramGasMap'

lorenzogentile404 committed:
added meaningful input for call to POINT_EVALUATION precompile in SixtyThreeSixtyFourthsPrecompileTests

OlivierBBB committed:
fix: initialize KZG

OlivierBBB committed:
constraints update

OlivierBBB committed:
ras: TracerTestBase now on OSAKA

OlivierBBB committed:
review comments

amkCha commented on arithmetization/src/main/java/net/consensys/linea/plugins/rpc/batchlinecount/ConflatedCountTracesV2.java:
```diff
@@ -117,7 +117,7 @@ private LineCountingTracer createLineCountingTracer(long fromBlock, long toBlock
 
     return tracerSharedConfiguration.isLimitless()
-        ? new ZkCounter(l1L2BridgeSharedConfiguration)
+        ? new ZkCounter(l1L2BridgeSharedConfiguration, fork)
```
This will require a PR in the monorepo as well to change the `ZkCounter` interface, everywhere `ZkCounter` is called https://github.com/Consensys/linea-monorepo/blob/5aed83a5f904a1b19906dac4aa0d45656074dfd3/besu-plugins/linea-sequencer/sequencer/src/main/java/net/consensys/linea/rpc/methods/LineaEstimateGas.java#L437 - might also be the case for Besu-shomei plugin

letypequividelespoubelles committed:
update constraints

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles reviewed (approved):
LGTM

letypequividelespoubelles committed:
Merge branch 'arith-dev' into EIP-7823-and-7883---Osaka-MODEXP-changes
